crates/re_viewer/src/ui/space_view_space_origin_ui.rs

```diff
@@ -0,0 +1,226 @@
+    space_origin_string: &mut String,
+    entered_editing: bool,
+    space_view: &SpaceViewBlueprint,
+    selected_suggestion: &mut Option<usize>,
+) -> bool {
+    let mut keep_editing = true;
+
+    //
+    // Build and filter the suggestion lists
+    //
+
+    // All suggestions for this class of space views.
+    let space_view_suggestions =
+        re_viewport::space_view_heuristics::all_possible_space_views(ctx, spaces_info)
+            .into_iter()
+            .map(|(space_view, _)| space_view)
+            .filter(|this_space_view| {
+                this_space_view.class_identifier() == space_view.class_identifier()
+            })
+            .collect::<Vec<_>>();
+
+    // Filtered suggestions based on the current text edit content.
+    let filtered_space_view_suggestions = space_view_suggestions
+        .iter()
+        .filter(|suggested_space_view| {
+            suggested_space_view
+                .space_origin
+                .to_string()
+                .contains(&*space_origin_string)
+        })
+        .collect::<Vec<_>>();
+
+    //
+    // Move cursor with keyboard (must happen before text edit to capture the keystrokes
+    //
+
+    let mut arrow_down =
+        ui.input_mut(|i| i.count_and_consume_key(Default::default(), Key::ArrowDown));
+    let arrow_up = ui.input_mut(|i| i.count_and_consume_key(Default::default(), Key::ArrowUp));
+
+    // force spawn a selected suggestion if the down arrow is pressed
+    if arrow_down > 0 && selected_suggestion.is_none() {
+        *selected_suggestion = Some(0);
+        arrow_down -= 1;
+    }
+
+    *selected_suggestion = selected_suggestion.map(|mut selected_suggestion| {
+        selected_suggestion = selected_suggestion
+            .saturating_add(arrow_down)
+            .saturating_sub(arrow_up);
+        if !space_view_suggestions.is_empty() {
+            selected_suggestion =
+                selected_suggestion.at_most(filtered_space_view_suggestions.len() - 1);
+        }
+        selected_suggestion
+    });
+
+    //
+    // Handle enter key when a suggestion is selected
+    //
+
+    let enter_key_hit = ui.input(|i| i.key_pressed(egui::Key::Enter));
+
+    if let Some(selected_suggestion) = selected_suggestion {
+        if enter_key_hit {
+            *space_origin_string = filtered_space_view_suggestions[*selected_suggestion]
+                .space_origin
+                .to_string();
+            keep_editing = false;
+        }
+    }
+
+    //
+    // Draw the text edit
+    //
+
+    let mut output = egui::TextEdit::singleline(space_origin_string).show(ui);
+
+    if entered_editing {
+        output.response.request_focus();
+        let min = egui::text::CCursor::new(0);
+        let max = egui::text::CCursor::new(space_origin_string.len());
+        let new_range = egui::text::CCursorRange::two(min, max);
+        output.state.cursor.set_char_range(Some(new_range));
+        output.state.store(ui.ctx(), output.response.id);
+    }
```

Comment:
@emilk I have an problem with this code. It should select _all_ the text, but if the click happens on the text, then it selects only all the way to where the mouse cursor is. I can't make sense of that behaviour. Any idea what's going on?

![Export-1706033974102](https://github.com/rerun-io/rerun/assets/49431240/60f0dffa-bb27-4282-8ca3-593c5487e69f)

Reply:
Weird - no I don't know

Reply:
issue with repro here: https://github.com/emilk/egui/issues/3883